saleor/payment/utils.py

```diff
@@ -427,3 +430,55 @@ def price_to_minor_unit(value: Decimal, currency: str):
     number_places = Decimal("10.0") ** precision
     value_without_comma = value * number_places
     return str(value_without_comma.quantize(Decimal("1")))
+
+
+def get_unfinished_payments():
+    """Return payments older than specified time and without assigned order object."""
+    ttl = settings.UNFINISHED_PAYMENT_TTL
+    now = timezone.now()
+    day_before = now - ttl
+
+    newest = (
+        Transaction.objects.filter(payment=OuterRef("pk"))
+        .filter(
+            kind__in=[TransactionKind.AUTH, TransactionKind.CAPTURE],
+            is_success=True,
+            action_required=False,
+        )
+        .order_by("-created")
+    )
+    payments = (
+        Payment.objects.get_queryset()
+        .filter(is_active=True, order=None)
+        .annotate(newest_trx_date=Subquery(newest.values("created")[:1]))
+        .filter(newest_trx_date__lte=day_before)
+    )
+
+    return payments
+
+
+def is_payment_unfinished_and_ready_to_release(payment: Payment):
+    """Payments older than specified time and without assigned order object."""
+    return get_unfinished_payments().filter(pk=payment.pk).exists()
+
+
+class ReleasePaymentException(Exception):
+    """Exception occured on attempt to release payment."""
+
+    pass
+
+
+def release_checkout_payment(payment: Payment, manager: "PluginsManager"):
+    """Try release payment calling refund or void."""
+    if not is_payment_unfinished_and_ready_to_release(payment):
```

Comment:
How many db queries this function will call? 
Also, I think that a raise msg is not correct.  Payment can not be ready to release, but it can be already relesed, or it can be assigned to Order.

Reply:
1. Jedno zapytanie per `payment` It calls `get_unfinished_payments` which create queryset and filter by payment.pk
```
SELECT ...
  (SELECT U0."created"
   FROM "payment_transaction" U0
   WHERE (U0."payment_id" = "payment_payment"."id"
          AND NOT U0."action_required"
          AND U0."is_success"
          AND U0."kind" IN (auth,
                            capture))
   ORDER BY U0."created" DESC
   LIMIT 1) AS "newest_trx_date"
FROM "payment_payment"
WHERE ("payment_payment"."is_active"
       AND "payment_payment"."order_id" IS NULL
       AND
         (SELECT U0."created"
          FROM "payment_transaction" U0
          WHERE (U0."payment_id" = "payment_payment"."id"
                 AND NOT U0."action_required"
                 AND U0."is_success"
                 AND U0."kind" IN (auth,
                                   capture))
          ORDER BY U0."created" DESC
          LIMIT 1) <= 2021-09-05 10:43:31.614135+00:00)
ORDER BY "payment_payment"."id" ASC
```

Reply:
2. As above. It calls `get_unfinished_payments` which consider only payments being active and without order. See
```
.filter(is_active=True, order=None)
```

Reply:
Wouldn't it be enough to do:

```
Payment.objects.filter(
    is_active=True,
    charge_status__in=[AUTHORIZED, CAPTURED],
    modified__lt=now() - TTL,
)
```

`gateway_postprocess` is already denormalizing results of `Transaction`s for us and storing it in a `Payment` object

Reply:
We can use a `payment.modified` but we need to make sure that we update it in all places where we operate on payment. 
Also, the proposed query probably needs to be modified to handle a Payment which we tried to refund/void but without any success.